zenith_utils/src/sync.rs

```diff
@@ -0,0 +1,133 @@
+use pin_project_lite::pin_project;
+use std::future::Future;
+use std::pin::Pin;
+use std::{io, task};
+
+pin_project! {
+    /// We use this future to mark certain methods
+    /// as callable in both sync and async modes
+    #[repr(transparent)]
+    pub struct SyncFuture<S, T: Future> {
+        #[pin]
+        nested: T,
+        _marker: std::marker::PhantomData<S>,
+    }
+}
+
+impl<S, T: Future> SyncFuture<S, T> {
+    pub(crate) fn new(nested: T) -> Self {
+        Self {
+            nested,
+            _marker: Default::default(),
+        }
+    }
+}
+
+impl<S, T: Future> Future for SyncFuture<S, T> {
+    type Output = T::Output;
+
+    #[inline(always)]
+    fn poll(self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> task::Poll<Self::Output> {
+        self.project().nested.poll(cx)
+    }
+}
+
+/// Proves that we can call [SyncFuture::wait]
+pub trait SyncProof {}
+
+impl<T: SyncProof> SyncProof for &T {}
+impl<T: SyncProof> SyncProof for &mut T {}
+
+impl<P: SyncProof, T: Future> SyncFuture<P, T> {
+    /// Synchronously wait for future completion
+    pub fn wait(mut self) -> T::Output {
+        const RAW_WAKER: task::RawWaker = task::RawWaker::new(
+            std::ptr::null(),
+            &task::RawWakerVTable::new(
+                |_| RAW_WAKER,
```

Comment:
Out of curiosity, how's what working in Rust?

I seem to fully miss the point on how it's possible to

1. call the constructor function as `const RAW_WAKER: task::RawWaker = task::RawWaker::new(`
2. create one of the function arguments to the constructor function as `&task::RawWakerVTable::new(`
3. reference itself from `1.` in `2.` as `|_| RAW_WAKER` in the argument creation in a `Fn` closure that's coerced to `unsafe fn(*const ()) -> RawWaker`

Syntactically looks like a circular dependency, where one object cannot really cbe constructed without another unless a compiler does some trick?

Also, there are no `Clone` or `Copy` impls for `RawWaker`, how that closure can exist or be called when we have moved the `RAW_WAKER` later into `task::Waker::from_raw`?

Reply:
FWIW, even the most trivial cycle is allowed syntactically (it obviously fails semantic checks, though):

```rust
error[E0391]: cycle detected when simplifying constant for the type system `FOO`
 --> const.rs:1:1
  |
1 | const FOO: u32 = FOO;
  | ^^^^^^^^^^^^^^^^^^^^^
  |
note: ...which requires simplifying constant for the type system `FOO`...
 --> const.rs:1:1
  |
1 | const FOO: u32 = FOO;
  | ^^^^^^^^^^^^^^^^^^^^^
note: ...which requires const-evaluating + checking `FOO`...
 --> const.rs:1:1
  |
1 | const FOO: u32 = FOO;
  | ^^^^^^^^^^^^^^^^^^^^^
  = note: ...which requires normalizing `FOO`...
  = note: ...which again requires simplifying constant for the type system `FOO`, completing the cycle
  = note: cycle used when running analysis passes on this crate

error: aborting due to previous error; 1 warning emitted
```

> Also, there are no Clone or Copy impls for RawWaker, how that closure can exist or be called when we have moved the RAW_WAKER later into task::Waker::from_raw

We haven't really moved anything. Consts are expanded each time we reference them, so it's not "that very same object".

I think it's not much different from recursive functions, where you can reference the function you haven't really defined yet. Here you have two mutually recursive thingies: a closure (function, really) and a value.

Reply:
Oh, I've missed const context entirely here, that explains things more.

Thanks for elaborating.